README.md
# whats_new

Beautiful way to showcase new features of your app.

[![pub package](https://img.shields.io/pub/v/whats_new.svg)](https://pub.dartlang.org/packages/whats_new)

![](https://media.giphy.com/media/7zutHtagBM6rlTknu2/giphy.gif)

## Getting Started

Add the dependency to your `pubspec.yaml`

```yaml
dependencies:
 whats_new: ^0.0.4
```

And run `flutter packages upgrade`

## How to use

### Option 1 

Use `WhatsNew.show(BuildContext, WhatsNewContent)`.

````dart
// Only show partial configurations, please refer WhatsNewContent to view more configurations
WhatsNew.show(context, WhatsNewContent(
      backgroundColor: Colors.white,
      title: 'What\'s New',
      titleColor: Colors.black,
      primaryButtonText: 'Let\'s Go',
      primaryButtonTextColor: Colors.white,
      primaryButtonBackgroundColor: Colors.red,
      secondaryButtonText: 'Learn more',
      secondaryButtonTextColor: Colors.red,
      transitionDuration: Duration(milliseconds: 350),
      featureItemTransitionDuration: Duration(milliseconds: 350),
      hideSecondaryButton: false,
      dismissible: false,
      callback: this,
      features: [
        Feature(
            icon: 'assets/camera-iris.png',
            iconColor: Colors.red,
            title: 'Title four',
            titleColor: Colors.black,
            description: 'Lorem ipsum dolor sit amet',
            descriptionColor: Colors.grey
        ),
        Feature(
            icon: 'assets/cellphone.png',
            iconColor: Colors.red,
            title: 'Title five',
            titleColor: Colors.black,
            description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt',
            descriptionColor: Colors.grey
        )
      ],
    ));
````
or
### Option 2

Created your own `WhatsNewContent` and launch with your own `Navigator`.

````dart
var content = WhatsNewContent(...)

Navigator.of(context).push(WhatsNew(content));
````

## Example

Check out the [example](https://pub.dartlang.org/packages/whats_new#-example-tab-) for more details.

## Changelog

See [changelog](https://pub.dartlang.org/packages/whats_new#-changelog-tab-).

## Contributions

Any contribution is more than welcome! You can contribute through pull requests and issues on GitHub.


CHANGELOG.md
## [1.0.0] - 2019-2-18

* Updated sdk to min 2.1.0
* Code cleanup

## [0.0.4] - 2018-10-25

* Add support to customise `fontSize` and `fontWeight`

## [0.0.3] - 2018-10-24

* Initial release


lib/util.dart
import 'package:flutter/material.dart';

class DisableGlowScrollBehavior extends ScrollBehavior {
  @override
  Widget buildViewportChrome(
      BuildContext context, Widget child, AxisDirection axisDirection) {
    return child;
  }
}

/// Keeps a Dart List in sync with an AnimatedList.
///
/// The [insert] and [removeAt] methods apply to both the internal list and the
/// animated list that belongs to [listKey].
///
/// This class only exposes as much of the Dart List API as is needed by the
/// sample app. More list methods are easily added, however methods that mutate the
/// list must make the same changes to the animated list in terms of
/// [AnimatedListState.insertItem] and [AnimatedList.removeItem].
class ListModel<E> {
  ListModel({
    @required this.listKey,
    @required this.removedItemBuilder,
    Iterable<E> initialItems,
  })  : assert(listKey != null),
        _items = List<E>.from(initialItems ?? <E>[]);

  final GlobalKey<AnimatedListState> listKey;
  final dynamic removedItemBuilder;
  final List<E> _items;

  AnimatedListState get _animatedList => listKey.currentState;

  void insert(int index, E item) {
    _items.insert(index, item);
    _animatedList.insertItem(index);
  }

  E removeAt(int index) {
    final E removedItem = _items.removeAt(index);
    if (removedItem != null) {
      _animatedList.removeItem(index,
          (BuildContext context, Animation<double> animation) {
        return removedItemBuilder(removedItem, context, animation);
      });
    }
    return removedItem;
  }

  int get length => _items.length;

  E operator [](int index) => _items[index];

  int indexOf(E item) => _items.indexOf(item);
}


lib/whats_new.dart
import 'package:flutter/material.dart';
import 'package:whats_new/whats_new_content.dart';

class WhatsNew extends ModalRoute<void> {
  static void show(BuildContext context, WhatsNewContent content) {
    Navigator.of(context).push(WhatsNew(content));
  }

  final WhatsNewContent content;

  WhatsNew(this.content) : assert(content != null);

  @override
  Duration get transitionDuration => this.content.transitionDuration;

  @override
  bool get opaque => false;

  @override
  bool get barrierDismissible => this.content.dismissible;

  @override
  Color get barrierColor => this.content.backgroundColor;

  @override
  String get barrierLabel => null;

  @override
  bool get maintainState => true;

  @override
  Widget buildPage(BuildContext context, Animation<double> animation,
      Animation<double> secondaryAnimation) {
    // This makes sure that text and other content follows the material style
    return Material(
      type: MaterialType.transparency,
      // make sure that the overlay content is not cut off
      child: SafeArea(
        child: this.content,
      ),
    );
  }

  @override
  Widget buildTransitions(BuildContext context, Animation<double> animation,
      Animation<double> secondaryAnimation, Widget child) {
    // You can add your own animations for the overlay content
    return FadeTransition(
      opacity: animation,
      child: child,
    );
  }
}


lib/whats_new_content.dart
import 'package:flutter/material.dart';
import 'package:whats_new/util.dart';

abstract class WhatsNewCallback {
  void onWhatsNewShowed();
  void onWhatsNewDismissed();
  void onPrimaryButtonClicked();
  void onSecondaryButtonClicked();
}

class Feature {
  final String icon;
  final Color iconColor;
  final String title;
  final Color titleColor;
  final double titleFontSize;
  final FontWeight titleFontWeight;
  final String description;
  final Color descriptionColor;
  final double descriptionFontSize;
  final FontWeight descriptionFontWeight;

  Feature({
    @required this.icon,
    @required this.iconColor,
    @required this.title,
    @required this.titleColor,
    @required this.description,
    @required this.descriptionColor,
    this.titleFontSize = 16.0,
    this.titleFontWeight = FontWeight.w600,
    this.descriptionFontSize = 16.0,
    this.descriptionFontWeight = FontWeight.normal,
  }) : assert(icon != null),
       assert(iconColor != null),
       assert(title != null),
       assert(titleColor != null),
       assert(titleFontSize != null),
       assert(titleFontWeight != null),
       assert(description != null),
       assert(descriptionColor != null),
       assert(descriptionFontSize != null),
       assert(descriptionFontWeight != null);
}

class WhatsNewContent extends StatefulWidget {
  final Color backgroundColor;
  final String title;
  final Color titleColor;
  final double titleFontSize;
  final FontWeight titleFontWeight;
  final String primaryButtonText;
  final Color primaryButtonTextColor;
  final Color primaryButtonBackgroundColor;
  final double primaryButtonFontSize;
  final FontWeight primaryButtonFontWeight;
  final String secondaryButtonText;
  final Color secondaryButtonTextColor;
  final double secondaryButtonFontSize;
  final FontWeight secondaryButtonFontWeight;
  final bool hideSecondaryButton;
  final WhatsNewCallback callback;
  final Duration transitionDuration;
  final Duration featureItemTransitionDuration;
  final bool dismissible;
  final List<Feature> features;

  WhatsNewContent({
    @required this.backgroundColor,
    @required this.title,
    @required this.titleColor,
    @required this.primaryButtonText,
    @required this.primaryButtonTextColor,
    @required this.primaryButtonBackgroundColor,
    @required this.secondaryButtonText,
    @required this.secondaryButtonTextColor,
    @required this.features,
    this.titleFontSize = 24.0,
    this.titleFontWeight = FontWeight.w600,
    this.primaryButtonFontSize = 18.0,
    this.primaryButtonFontWeight = FontWeight.normal,
    this.secondaryButtonFontSize = 16.0,
    this.secondaryButtonFontWeight = FontWeight.normal,
    this.hideSecondaryButton = false,
    this.callback,
    this.transitionDuration = const Duration(milliseconds: 350),
    this.featureItemTransitionDuration = const Duration(milliseconds: 350),
    this.dismissible = false,
  })  : assert(backgroundColor != null),
        assert(title != null),
        assert(titleColor != null),
        assert(titleFontSize != null),
        assert(titleFontWeight != null),
        assert(primaryButtonText != null),
        assert(primaryButtonTextColor != null),
        assert(primaryButtonBackgroundColor != null),
        assert(primaryButtonFontSize != null),
        assert(primaryButtonFontWeight != null),
        assert(secondaryButtonText != null),
        assert(secondaryButtonTextColor != null),
        assert(secondaryButtonFontSize != null),
        assert(secondaryButtonFontWeight != null),
        assert(features != null),
        assert(hideSecondaryButton != null),
        assert(transitionDuration != null),
        assert(dismissible != null);

  @override
  WhatsNewContentState createState() {
    return new WhatsNewContentState();
  }
}

class WhatsNewContentState extends State<WhatsNewContent> {
  final GlobalKey<AnimatedListState> _listKey = GlobalKey<AnimatedListState>();
  ListModel<Feature> _features;

  @override
  void initState() {
    super.initState();
    this._features = ListModel<Feature>(
      listKey: _listKey,
      initialItems: [],
      removedItemBuilder: null,
    );
  }

  @override
  Widget build(BuildContext context) {
    this.widget.callback?.onWhatsNewShowed();

    List<Widget> contents = [];
    contents.add(this._titleLayout(context));
    contents.add(this._featuresLayout(context));
    if (!this.widget.hideSecondaryButton) {
      contents.add(this._actionsLayout(context));
    }

    if (this.widget.features.length > 0) {
      this.iterateAndAddFeatures(0);
    }

    return new Scaffold(
      body: new Container(
        color: this.widget.backgroundColor,
        child: new SafeArea(
            child: new Container(
          padding: EdgeInsets.all(16.0),
          child: new Column(
            mainAxisAlignment: MainAxisAlignment.spaceBetween,
            children: contents,
          ),
        )),
      ),
    );
  }

  void iterateAndAddFeatures(int index) async {
    if (index <= this.widget.features.length - 1) {
      Future.delayed(this.widget.featureItemTransitionDuration, () {
        this
            ._features
            .insert(this._features.length, this.widget.features[index]);
        this.iterateAndAddFeatures(index + 1);
      });
    }
  }

  Widget _titleLayout(BuildContext context) {
    return new Container(
      alignment: Alignment.center,
      padding: EdgeInsets.only(top: 16.0, bottom: 16.0),
      child: Text(
        this.widget.title,
        style: Theme.of(context).textTheme.display1.copyWith(
          color: this.widget.titleColor,
          fontWeight: this.widget.titleFontWeight,
          fontSize: this.widget.titleFontSize,
        ),
        maxLines: 3,
      ),
    );
  }

  Widget _featuresLayout(BuildContext context) {
    return new Expanded(
        flex: 4,
        child: new Container(
          margin:
              EdgeInsets.only(left: 4.0, top: 16.0, right: 4.0, bottom: 16.0),
          child: new ScrollConfiguration(
            behavior: DisableGlowScrollBehavior(),
            child: AnimatedList(
                key: this._listKey,
                initialItemCount: this._features.length,
                itemBuilder: (context, index, animation) {
                  return this._featureLayout(
                      context, this._features[index], animation);
                }),
          ),
        ));
  }

  Widget _featureLayout(
      BuildContext context, Feature feature, Animation<double> animation) {
    return new Container(
      child: SizeTransition(
        sizeFactor: animation,
        child: ListTile(
          isThreeLine: true,
          enabled: false,
          leading: Image.asset(
            feature.icon,
            color: feature.iconColor,
            width: 40.0,
            height: 40.0,
          ),
          title: Text(
            feature.title,
            style: Theme.of(context).textTheme.subhead.copyWith(
              color: feature.titleColor,
              fontSize: feature.titleFontSize,
              fontWeight: feature.titleFontWeight,
            ),
            maxLines: 2,
          ),
          subtitle: Text(
            feature.description,
            style: Theme.of(context).textTheme.subhead.copyWith(
              color: feature.descriptionColor,
              fontSize: feature.descriptionFontSize,
              fontWeight: feature.descriptionFontWeight,
            ),
            maxLines: 5,
          ),
        ),
      ),
    );
  }

  Widget _actionsLayout(BuildContext context) {
    return new Container(
      alignment: Alignment.bottomCenter,
      child: new Column(
    mainAxisSize: MainAxisSize.min,
    children: <Widget>[
      new Container(
        padding:
            EdgeInsets.only(left: 8.0, top: 4.0, right: 8.0, bottom: 4.0),
        margin: EdgeInsets.only(bottom: 16.0),
        child: GestureDetector(
          onTap: () {
            this.widget.callback?.onSecondaryButtonClicked();
          },
          child: Text(
            this.widget.secondaryButtonText,
            style: Theme.of(context).textTheme.body1.copyWith(
              color: this.widget.secondaryButtonTextColor,
              fontSize: this.widget.secondaryButtonFontSize,
              fontWeight: this.widget.secondaryButtonFontWeight
            ),
          ),
        ),
      ),
      SizedBox(
        width: MediaQuery.of(context).size.width,
        child: RaisedButton(
          onPressed: () {
            this.widget.callback?.onPrimaryButtonClicked();
            this.widget.callback?.onWhatsNewDismissed();
            Navigator.of(context).pop();
          },
          child: Text(
            this.widget.primaryButtonText,
            style: Theme.of(context).textTheme.subhead.copyWith(
              color: this.widget.primaryButtonTextColor,
              fontSize: this.widget.primaryButtonFontSize,
              fontWeight: this.widget.primaryButtonFontWeight
            ),
          ),
          shape: RoundedRectangleBorder(
              side: BorderSide(
                  color: this.widget.primaryButtonBackgroundColor),
              borderRadius: BorderRadius.all(Radius.circular(10.0))
          ),
          color: this.widget.primaryButtonBackgroundColor,
          padding: EdgeInsets.only(top: 20.0, bottom: 20.0),
        ),
      ),
    ],
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:whats_new/whats_new.dart';
import 'package:whats_new/whats_new_content.dart';

// Example application:
void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Playground',
      home: TestPage(),
    );
  }
}

class TestPage extends StatelessWidget implements WhatsNewCallback {
  void _showOverlay(BuildContext context) {
    WhatsNew.show(context, WhatsNewContent(
      backgroundColor: Colors.white,
      title: 'What\'s New',
      titleColor: Colors.black,
      primaryButtonText: 'Let\'s Go',
      primaryButtonTextColor: Colors.white,
      primaryButtonBackgroundColor: Colors.red,
      secondaryButtonText: 'Learn more',
      secondaryButtonTextColor: Colors.red,
      transitionDuration: Duration(milliseconds: 350),
      featureItemTransitionDuration: Duration(milliseconds: 350),
      hideSecondaryButton: false,
      dismissible: false,
      callback: this,
      features: [
        Feature(
          icon: 'assets/access-point.png',
          iconColor: Colors.red,
          title: 'Title one',
          titleColor: Colors.black,
          description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.',
          descriptionColor: Colors.grey
        ),
        Feature(
            icon: 'assets/account.png',
            iconColor: Colors.red,
            title: 'Title two',
            titleColor: Colors.black,
            description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat.',
            descriptionColor: Colors.grey
        ),
        Feature(
            icon: 'assets/alert-circle.png',
            iconColor: Colors.red,
            title: 'Title three',
            titleColor: Colors.black,
            description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit',
            descriptionColor: Colors.grey
        ),
        Feature(
            icon: 'assets/camera-iris.png',
            iconColor: Colors.red,
            title: 'Title four',
            titleColor: Colors.black,
            description: 'Lorem ipsum dolor sit amet',
            descriptionColor: Colors.grey
        ),
        Feature(
            icon: 'assets/cellphone.png',
            iconColor: Colors.red,
            title: 'Title five',
            titleColor: Colors.black,
            description: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt',
            descriptionColor: Colors.grey
        )
      ],
    ));
  }

  @override
  void onWhatsNewShowed() {
    print('onWhatsNewShowed');
  }

  @override
  void onWhatsNewDismissed() {
    print('onWhatsNewDismissed');
  }

  @override
  void onPrimaryButtonClicked() {
    print('onPrimaryButtonClicked');
  }

  @override
  void onSecondaryButtonClicked() {
    print('onSecondaryButtonClicked');
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: Text('flutter_whats_new_example')),
      body: Padding(
        padding: EdgeInsets.all(16.0),
        child: Center(
          child: RaisedButton(
            onPressed: () => _showOverlay(context),
            child: Text('Show WhatsNew'),
          ),
        ),
      ),
    );
  }
}